ffi/capi/src/datetime_helpers.rs

```diff
@@ -0,0 +1,227 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#[cfg(any(feature = "compiled_data", feature = "buffer_provider"))]
+use alloc::boxed::Box;
+#[cfg(any(feature = "compiled_data", feature = "buffer_provider"))]
+use icu_calendar::Gregorian;
+#[cfg(any(feature = "compiled_data", feature = "buffer_provider"))]
+use icu_datetime::{
+    fieldsets::builder::BuilderError, fieldsets::enums::*, fieldsets::Combo, pattern::*,
+    scaffold::*, DateTimeFormatter, DateTimeFormatterLoadError, FixedCalendarDateTimeFormatter,
+};
+
+#[cfg(any(feature = "compiled_data", feature = "buffer_provider"))]
+pub(crate) fn map_or_default<Input, Output>(input: Option<Input>) -> Output
+where
+    Output: From<Input> + Default,
+{
+    input.map(Output::from).unwrap_or_default()
+}
+
+#[cfg(any(feature = "compiled_data", feature = "buffer_provider"))]
+pub(super) fn date_formatter_with_zone<Zone>(
+    formatter: &DateTimeFormatter<DateFieldSet>,
+    locale: &crate::locale_core::ffi::Locale,
+    zone: Zone,
+    load: impl FnOnce(
+        &mut DateTimeNames<Combo<DateFieldSet, Zone>>,
+    ) -> Result<(), crate::errors::ffi::DateTimeFormatterLoadError>,
+    to_formatter: impl FnOnce(
+        DateTimeNames<Combo<DateFieldSet, Zone>>,
+        Combo<DateFieldSet, Zone>,
+    ) -> Result<
+        DateTimeFormatter<Combo<DateFieldSet, Zone>>,
```

Comment:
nit: we should use some typedefs to simplify this, or skip rustfmt here so that stuff is mostly on the same line, the line splitting makes this really hard to understand what's going on here.